cmake: toolchain/xt-clang: allow to skip FLIX generation

dcpleung opened the pull request:
This adds compiler flag to instruct xt-clang not to generate FLIX (Flexible Length Instruction Xtensions) instructions, where multiple instructions can be fused into one instruction. FLIX can provide performance and code size advantages.

When userspace is enabled, FLIX may reorder memory access where memory is accessed before code determines whether we should perform that memory access. For example, when guarding memory access via k_is_user_context(), and some kernel variables can only be accessed via kernel mode, FLIX may reorder memory access such that these variables are accessed before k_is_user_context(). This results in access violation.

FLIX generation is enabled by default. This lets the application developers to disable FLIX generation if needed.

tejlmand commented on arch/xtensa/core/CMakeLists.txt:
```diff
@@ -2,6 +2,10 @@
 zephyr_cc_option(-mlongcalls)
 
+if(CONFIG_TOOLCHAIN_XTCLANG_DO_NOT_GENERATE_FLIX)
+  zephyr_cc_option(-mno-generate-flix)
```
please add this flag to https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/compiler/xt-clang/compiler_flags.cmake and give it a proper name.

We try to avoid, or at least minimize, the direct use of compiler flags as it makes it harder to support multiple toolchains.
I do notice this resides in `arch/xtensa` but principle easily gets copied elsewhere.

tejlmand commented on cmake/toolchain/xt-clang/Kconfig:
```diff
@@ -0,0 +1,22 @@
+config TOOLCHAIN_XTCLANG_DO_NOT_GENERATE_FLIX
+	bool "Skip instruction fusions via FLIX"
+	depends on DT_HAS_CDNS_TENSILICA_XTENSA_LX7_ENABLED
+	default y if USERSPACE
```
Please turn this setting around so it is disabled per default, and then can be enabled by samples which wants it.
Like:
```suggestion
config TOOLCHAIN_XTCLANG_FLIX
	bool "Flexible Length Instruction Xtensions (FLIX)"
	depends on DT_HAS_CDNS_TENSILICA_XTENSA_LX7_ENABLED
	default y if !USERSPACE
```

Related comment, but not a required change.
Afaik FLIX is just Xtensa's form of `VLIW`, so a more generic option would be to relate the feature to `VLIW` and then let xt-clang define this to be `-mno-generate-flix`.

dcpleung committed:
cmake: compiler: add flags for VLIW generations

This adds to compiler flags template on instructing the compiler
on whether to generate VLIW instructions.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
cmake: toolchain/xt-clang: add flag to skip FLIX generation

This adds the compiler flag to instruct xt-clang not to generate
FLIX (Flexible Length Instruction Xtensions) instructions, where
multiple instructions can be fused into one instruction. FLIX
can provide performance and code size advantages.

When userspace is enabled, FLIX may reorder memory access
where memory is accessed before code determines whether
we should perform that memory access. For example,
when guarding memory access via k_is_user_context(), and
some kernel variables can only be accessed via kernel mode,
FLIX may reorder memory access such that these variables are
accessed before k_is_user_context(). This results in access
violation. This turns off FLIX generation by default and
lets the application developers to decide whether to enable
FLIX generation.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Modified to do the compiler flag route.

tejlmand reviewed (dismissed):
looks good.

A smaller, non-blocking, suggestion to consider.

lyakh commented on cmake/toolchain/xt-clang/Kconfig:
```diff
@@ -0,0 +1,11 @@
+
+endchoice
+
+endif # DT_HAS_CDNS_TENSILICA_XTENSA_LX7_ENABLED
```
should this also be converted? https://github.com/zephyrproject-rtos/zephyr/blob/c86c748ff037259412edb6f3a6bea52a7efad87c/cmake/compiler/xt-clang/target.cmake#L31

lyakh replied:
@dcpleung I mean - should that also use the new compiler property?

softwarecki commented:
I assume you have already tested different approaches, but I want to make sure. I don't like the workaround that disables optimization for the entire project. In my opinion, this is too intrusive. Wouldn't it be better to apply a smaller change only in the problematic spots? I assume the issue with early access to the kernel-only variable occurs in very few places.

Have you tried moving the declaration of this variable below the if? Another option could be defining a pointer to this variable and using it in the return. As a last resort, you could separate access to this variable into a separate function that would be called during return.

The most important question:
Have you tried other barriers? In `include/zephyr/sys/barrier_builtin.h` there are three types of barriers defined. Only `z_barrier_isync_fence_full` uses the assembler `memory` instruction. The other two, `z_barrier_dmem_fence_full` and `z_barrier_dsync_fence_full`, use gcc built-in function `__atomic_thread_fence`.
The `compiler_barrier` also uses the `memory` instruction, which is the `isync` variant. Maybe if we inform the compiler that we want a barrier at this point, it would generate the code differently? Perhaps the `memory` instruction is not clear enough for it? I think looking at the assembly code of this function could help us understand the problem.

lyakh commented:
@dcpleung are there any xtensa asm instructions that are never included in FLIX?

wjablon1 commented:
If the issue occurs in multiple places, there is a risk that we miss some of them by applying a manual approach like barriers. If we sprinkle the entire kernel code with those barriers, the performance could be the same as with FLIX disabled (or even worse) while the code readability decreased.

The performance measurements would ideal. Also, it’s worth figuring out why Xtensa XOS has just mno-l32r-flix. Maybe there are some special programming techniques used or a combination of compiler options.

I partly agree with @softwarecki that the solution is too intrusive. Some algorithms, that have zero interactions with the kernel, vastly benefit from FLIX. But as a temporary solution, I guess it’s not the end of the world. Especially because FLIX is now disabled for LLEXT modules for another reason (linking problem).

wjablon1 commented on Kconfig.zephyr:
```diff
@@ -629,6 +629,37 @@ config TOOLCHAIN_SUPPORTS_VLA_IN_STATEMENTS
 	  Hidden symbol to state if the toolchain can handle vla in
 	  statements.
 
+menu "Code Generation"
```
Wouldn't be better to have more flexible/generic solution like in case of "config COMPILER_OPT" (or reuse that if possible)? 
My concern is that if someone wants to fine-tune other optimizations/instructions of a complier, her or she probably needs to follow the same pattern. You can imagine how the kconfig could bloat in the future.

There are more possibilities related to FLIX in the Xtensa compiler. So the solution isn't comprehensive from the beginning. Also I would like to add a similar option but only for LLEXT modules. Should I copy-paste the entire choice section?

softwarecki commented:
Could you share a part of the assembler code where the FLIX is generated incorrectly?

softwarecki commented:
> Besides, fixing one macro does not fix the fundamental issue where the compiler will generate this kind of pattern

Our responsibility is to tell the compiler where its optimizations must not apply. Using memory barriers in kernel code is not a workaround, it is the correct way to enforce ordering rules. For me the real workaround would be globally disabling the FLIX optimization in a case where the issue does not even occur in the FLIX instruction.

Please note that memory barriers are already used with `k_is_user_context` context. For example, the `ARCH_EXCEPT` macro on RISC-V, and `k_is_pre_kernel` use barriers to enforce ordering. The `gen_syscalls.py` script also inserts a barrier after `z_syscall_trap` (which calls `arch_is_user_context`).

There are projects where using macros is considered a bad practice. Here we clearly went in the opposite direction and crossed the critical mass of macros. At this point they are not a tool anymore, they are a lifestyle choice.

dcpleung commented:
Note that with FLIX generation disabled, the compiler emits code with correct ordering. And that no other compilers have this issue either. From my perspective, adding `compiler_barrier()` is a workaround for a misbehaving compiler... especially as the incorrect ordering is done even without emitting FLIX instructions.

Just looking at your PR, there are lots of code and complexities added for one compiler. And that's going to add quite some emitted code in binaries... for other compilers which absolutely do not need the extra baggage.

SzymonRichert commented on Kconfig.zephyr:
```diff
@@ -629,6 +629,37 @@ config TOOLCHAIN_SUPPORTS_VLA_IN_STATEMENTS
+	  the compiler to fuse instructions for parallel execution.
+
+config COMPILER_CODEGEN_VLIW_AUTO
+	bool "Let compiler decide"
```
@tejlmand can you take a look ?
